cryptography/src/bls12381/dkg.rs

```diff
@@ -0,0 +1,1702 @@
+                &weights,
+                round_info.threshold(),
+                concurrency,
+            );
+            if previous.public().constant() != public.constant() {
+                return Err(Error::DkgFailed);
+            }
+            (public, Some(weights))
+        } else {
+            let mut public = Poly::zero();
+            for (_, log) in selected.iter() {
+                public.add(&log.pub_msg.commitment);
+            }
+            (public, None)
+        };
+        let output = Output {
+            summary: round_info.summary,
+            players: round_info.players,
+            public,
+        };
+        Ok(Self { output, weights })
+    }
+}
+
+/// Observe the result of a DKG, using the public results.
+///
+/// The log mapping dealers to their log is the shared piece of information
+/// that the participants (players, observers) of the DKG must all agree on.
+///
+/// From this log, we can (potentially, as the DKG can fail) compute the public output.
+///
+/// This will only ever return [`Error::DkgFailed`].
+pub fn observe<V: Variant, P: PublicKey>(
+    round_info: Info<V, P>,
+    logs: BTreeMap<P, DealerLog<V, P>>,
+    concurrency: usize,
+) -> Result<Output<V, P>, Error> {
+    let selected = select(&round_info, logs)?;
+    ObserveInner::<V, P>::reckon(round_info, selected, concurrency).map(|x| x.output)
+}
+
+/// Represents a player in the DKG / reshare process.
+///
+/// The player is attempting to get a share of the key.
+///
+/// They need not have participated in prior rounds.
+pub struct Player<V: Variant, S: Signer> {
+    me: S,
+    me_pub: S::PublicKey,
+    round_info: Info<V, S::PublicKey>,
+    index: u32,
+    transcript: Transcript,
+    view: BTreeMap<S::PublicKey, (DealerPubMsg<V>, DealerPrivMsg)>,
+}
+
+impl<V: Variant, S: Signer> Player<V, S> {
+    /// Create a new [`Player`].
+    ///
+    /// We need the player's private key in order to sign messages.
+    pub fn new(round_info: Info<V, S::PublicKey>, me: S) -> Result<Self, Error> {
+        let me_pub = me.public_key();
+        Ok(Self {
+            index: round_info.player_index(&me_pub)?,
+            me,
+            me_pub,
+            transcript: transcript_for_round(&round_info),
+            round_info,
+            view: BTreeMap::new(),
+        })
+    }
+
+    /// Process a message from a dealer.
+    ///
+    /// It's important that nobody can impersonate the dealer, and that the
+    /// private message was not exposed to anyone else. A convenient way to
+    /// provide this is by using an authenticated channel, e.g. via
+    /// [crate::handshake], or [commonware-p2p](https://docs.rs/commonware-p2p/latest/commonware_p2p/).
+    pub fn dealer_message(
+        &mut self,
+        dealer: S::PublicKey,
+        pub_msg: DealerPubMsg<V>,
+        priv_msg: DealerPrivMsg,
+    ) -> Option<PlayerAck<S::PublicKey>> {
+        if self.view.contains_key(&dealer) {
+            return None;
```

Comment:
In the previous implementation, I used to delineate between `skip` and `failure`. It definitely makes it a lot easier to treat them the same.